bolt-modules/boltlib/lib/puppet/functions/run_plan.rb

```diff
@@ -85,6 +85,31 @@ def run_plan(scope, plan_name, named_args = {})
       executor.report_yaml_plan(closure.model.body)
     end
 
+    # If a TargetSpec parameter is passed, ensure it is in inventory
+    inventory = Puppet.lookup(:bolt_inventory)
+    # Is it possible that inventory is not here?
+    if inventory.version > 1
+      parameter_names_to_type = closure.parameters.each_with_object({}) do |param, param_acc|
+        # parameter has multiple keys ex Optional[TargetSpec]
+        param_acc[param.name] = if defined?(param.type_expr.keys)
+                                  param.type_expr.keys.each_with_object([]) do |param_type, type_acc|
+                                    type_acc << param_type.cased_value
+                                  end
+                                # Only a single key, may or may not have a type specified
+                                else
+                                  param.type_expr ? [param.type_expr.cased_value] : []
+                                end
+      end
+
+      params.each do |param, value|
+        # CODEREVIEW: is it safe to match on with i =~ /TargetSpec/ it handles the BoltLib::TargetSpec
+        # but would I guess could match on any type werhe cased_value matches TargetSpec
+        if parameter_names_to_type[param]&.any? { |i| i =~ /TargetSpec/ }
```

Comment:
I can't think of a reasonable type that would falsely match this (hopefully no one is doing `Variant[TargetSpec, Integer]`). I'm inclined to wait until it's a problem. With that said it might be safer to just match `TargetSpec` and `Optional[TargetSpec]`, since I expect that will be 99% of cases.

Reply:
I added some ugly logic to resolve type aliases. I think that I can just check for two conditions, `Boltlib::TargetSpec` and `TargetSpec`.